stage0/src/initramfs.rs

```diff
@@ -16,38 +16,135 @@
+        .expect("couldn't read initial RAM disk size") as usize;
     if size == 0 {
         log::debug!("No initial RAM disk");
         return None;
     }
 
-    let address = fw_cfg
+    let mut initrd_address = fw_cfg
         .read_initrd_address()
         .expect("couldn't read initial RAM disk address");
 
-    // We use an identity mapping for memory.
-    let address = VirtAddr::new(address.as_u64());
-
     log::debug!("Initial RAM disk size {}", size);
-    log::debug!("Initial RAM disk address {:#018x}", address.as_u64());
+    log::debug!("Initial RAM disk address {:#018x}", initrd_address.as_u64());
+
+    let kernel_address = fw_cfg
+        .read_kernel_address()
+        .expect("couldn't read kernel address");
+    let kernel_size = fw_cfg
+        .read_kernel_size()
+        .expect("couldn't read kernel address") as u64;
+
+    log::debug!("Kernel size {}", kernel_size);
+    log::debug!("Kernel start address {:#018x}", kernel_address.as_u64());
+
+    if let Err(error) = check_initrd_address(
+        initrd_address,
+        initrd_address + (size as u64),
+        kernel_address,
+        kernel_address + kernel_size,
+        e820_table,
+    ) {
+        log::warn!("Unsuitable initial RAM disk address: {}", error);
+        initrd_address = find_suitable_initrd_address(size, e820_table)
+            .expect("couldn't a find suitable initial RAM disk address");
+        log::debug!(
+            "Suggested new initial RAM disk address: {:#018x}",
+            initrd_address.as_u64()
+        );
+    }
+
+    // If it is still not a suitable address, just panic.
+    check_initrd_address(
+        initrd_address,
+        initrd_address + size as u64,
+        kernel_address,
+        kernel_address + kernel_size,
+        e820_table,
+    )
+    .expect("couldn't find a suitable initial RAM disk address");
 
-    // TODO(#3628): Check that the suggested address is valid and sensible.
-    // Safety: for now we trust that the hypervisor suggests a valid address and to load the RAM
-    // disk, but will add checks to ensure it falls in valid mapped memory and won't overwrite
-    // existing structures. We only write to the slice and don't assume anything about its layout.
-    let buf = unsafe { slice::from_raw_parts_mut::<u8>(address.as_mut_ptr(), size as usize) };
+    // We use an identity mapping for memory.
+    let address = VirtAddr::new(initrd_address.as_u64());
+
+    // Safety: We already checked that the slice falls within the mapped virtual memory, is backed
+    // by physical memory, does not overlap with the kernel, and does not overlap with the first
+    // 2MiB where our important data structures live. It can also not overlap with the firmware ROM
+    // image, since that is above the 1GiB upper limit we support. We only write to the slice and
+    // don't assume anything about its layout or alignment.
+    let buf = unsafe { slice::from_raw_parts_mut::<u8>(address.as_mut_ptr(), size) };
```

Comment:
We should probably mark the memory as reserved in the E820 map as well, for both (a) safety that we won't stomp on the data structure later, and (b) ensuring that the kernel itself doesn't accidentally overwrite the memory.

(Or, check what SeaBIOS and/or the linuxboot option ROMs are doing -- do they mark the memory as reserved?)

Reply:
They don't mark it as reserved, and I also think we should not. This is a temporary compressed archive. The kernel extracts it to somewhere else in memory before actually mounting it as a ramdisk.